packages/core/src/lib/light_push/light_push_v3.ts

```diff
@@ -0,0 +1,300 @@
+
+function isValidRequestId(requestId: string | undefined): boolean {
+  if (!requestId) return false;
+  if (requestId === "N/A") return true;
+  return UUID_REGEX.test(requestId);
+}
+
+const STREAM_TIMEOUT_MS = 30000;
+
+function withTimeout<T>(promise: Promise<T>, timeoutMs: number): Promise<T> {
+  return Promise.race([
+    promise,
+    new Promise<T>((_, reject) =>
+      setTimeout(
+        () => reject(new Error("Stream operation timed out")),
+        timeoutMs
+      )
+    )
+  ]);
+}
+
+type PreparePushMessageResult = ThisOrThat<"query", PushRpcV3>;
+
+export class LightPushCoreV3 extends BaseProtocol implements IBaseProtocolCore {
+  public constructor(
+    public readonly pubsubTopics: PubsubTopic[],
+    libp2p: Libp2p
+  ) {
+    super(LightPushCodecV3, libp2p.components, pubsubTopics);
+  }
+
+  private async preparePushMessage(
+    encoder: IEncoder,
+    message: IMessage
+  ): Promise<PreparePushMessageResult> {
+    try {
+      if (!message.payload || message.payload.length === 0) {
+        log.error("Failed to send waku light push: payload is empty");
+        return { query: null, error: ProtocolError.EMPTY_PAYLOAD };
+      }
+
+      if (!(await isMessageSizeUnderCap(encoder, message))) {
+        log.error("Failed to send waku light push: message is bigger than 1MB");
+        return { query: null, error: ProtocolError.SIZE_TOO_BIG };
+      }
+
+      const protoMessage = await encoder.toProtoObj(message);
+      if (!protoMessage) {
+        log.error("Failed to encode to protoMessage, aborting push");
+        return {
+          query: null,
+          error: ProtocolError.ENCODE_FAILED
+        };
+      }
+
+      const query = PushRpcV3.createRequest(
+        protoMessage as WakuMessage,
+        encoder.pubsubTopic
+      );
+      return { query, error: null };
+    } catch (error) {
+      log.error("Failed to prepare push message", error);
+
+      return {
+        query: null,
+        error: ProtocolError.GENERIC_FAIL
+      };
+    }
+  }
+
+  public async send(
+    encoder: IEncoder,
+    message: IMessage,
+    peerId: PeerId
+  ): Promise<CoreProtocolResultWithMeta> {
+    const { query, error: preparationError } = await this.preparePushMessage(
+      encoder,
+      message
+    );
+
+    if (preparationError || !query) {
+      return {
+        success: null,
+        failure: {
+          error: preparationError,
+          peerId
+        },
+        protocolUsed: LightPushCodecV3
+      };
+    }
+
+    let stream: Stream;
+    try {
+      stream = await this.getStream(peerId);
+    } catch (error) {
+      log.error("Failed to get stream", error);
+      return {
+        success: null,
+        failure: {
+          error: ProtocolError.NO_STREAM_AVAILABLE,
+          peerId: peerId
+        },
+        protocolUsed: LightPushCodecV3
+      };
+    }
+
+    let res: Uint8ArrayList[] | undefined;
+    try {
+      res = await withTimeout(
```

Comment:
in another place it is not implemented with timeout, I would keep it like that and create an issue to follow up